scripts/NuGetUtilities.psm1

```diff
@@ -0,0 +1,155 @@
+        "--include-source", "--include-symbols",
+        "-p:Platform=$Platform",
+        "--output", (Get-PackageDirectoryName $configuration)
+
+    Write-CommandLine dotnet $arguments
+
+    dotnet $arguments
+    if ($LASTEXITCODE -ne 0) {
+        Exit-WithFailureMessage $ScriptName "$project NuGet package creation failed."
+    }
+}
+
+function New-NuGetPackageFromNuSpecFile($configuration, $project, $version, $suffix = "") {
+    $nuspecFile = "$SourceRoot\NuGet\$project.nuspec"
+
+    $arguments=
+        "pack", $nuspecFile,
+        "-Symbols",
+        "-Properties", "platform=$Platform;configuration=$configuration;version=$version",
+        "-Verbosity", "Quiet",
+        "-BasePath", ".\",
+        "-OutputDirectory", (Get-PackageDirectoryName $configuration)
+
+    if ($suffix -ne "") {
+        $arguments += "-Suffix", $Suffix
+    }
+
+    Write-CommandLine $NuGetExePath $arguments
+
+    & $NuGetExePath $arguments
+    if ($LASTEXITCODE -ne 0) {
+        Exit-WithFailureMessage $ScriptName "$project NuGet package creation failed."
+    }
+
+    Write-Information "  Successfully created package '$BinRoot\NuGet\$Configuration\$Project.$version.nupkg'."
+}
+
+function New-NuGetPackages($configuration, $projects) {
+    $version = Get-PackageVersion
+
+    # We can build the NuGet packages for library projects directly from their
+    # project file.
+    foreach ($project in $projects.NewLibrary) {
+        New-NuGetPackageFromProjectFile $configuration $project $version
+    }
+
+    # Unfortunately, application projects like MultiTool need to include things
+    # that are not specified in the project file, so their packages still require
+    # a .nuspec file.
+    foreach ($project in $Projects.NewApplication) {
+        New-NuGetPackageFromNuSpecFile $configuration $project $version
+    }
+}
+
+function Get-NuGetApiKey {
+    # We temporarily implement this function by parsing the key from the file
+    # SetNuGetJSarifApiKey.cmd, which is expected to exist in the parent
+    # directory of the developer's sarif-sdk enlistment.
+    # In future, we should get this key from the Azure Key Vault.
+    $apiKeyPath = Join-Path $PSScriptRoot ..\..\SetNuGetSarifApiKey.cmd
+    if (-not (Test-Path $apiKeyPath)) {
+        Exit-WithFailureMessage NuGetUtilties "API key file $apiKeyPath does not exist."
+    }
+
+    # Everything that isn't a double quote, followed by a double quote, followed
+    # by the key (everything up to the next double quote).
+    $pattern = '^[^"]*"(?<key>[^"]*)'
+
+    $firstLine = [IO.File]::ReadAllLines($apiKeyPath)[0]
+    if ($firstLine -match $pattern) {
+        $matches["key"]
+    } else {
+        Exit-WithFailureMessage NuGetUtilties "API key file $apiKeyPath does not contain a key."
+    }
+}
+
+function Set-NuGetApiKey {
+    $apiKey = Get-NuGetApiKey
+
+    $arguments = "SetApiKey", $apiKey, "-Source", $PackageSource, "-Verbosity", "quiet"
+    Write-CommandLine $NuGetExePath $arguments
```

Comment:
>Write-CommandLine [](start = 4, length = 17)

We frequently echo the command line before executing the command. This utility function DRYs out the code that does that.

Reply:
We could DRY it out even further by bundling together the echo and the execution into a utility function `Run-CommandLine`.

---
In reply to: [209440001](https://github.com/Microsoft/sarif-sdk/pull/1010#discussion_r209440001) [](ancestors = 209440001)